analyses/cms-open-data-ttbar/ml.py

```diff
@@ -0,0 +1,297 @@
+        "p_T of the first W Jet [GeV]",
+        "p_T of the second W Jet [GeV]",
+        "p_T of the b_{top-had} Jet [GeV]",
+        "p_T of the b_{top-lep} Jet [GeV]",
+        "btagCSVV2 of the first W Jet",
+        "btagCSVV2 of the second W Jet",
+        "btagCSVV2 of the b_{top-had} Jet",
+        "btagCSVV2 of the b_{top-lep} Jet",
+        "Quark vs Gluon likelihood discriminator of the first W Jet",
+        "Quark vs Gluon likelihood discriminator of the second W Jet",
+        "Quark vs Gluon likelihood discriminator of the b_{top-had} Jet",
+        "Quark vs Gluon likelihood discriminator of the b_{top-lep} Jet",
+    ]
+}
+
+def define_cpp (fastforest_path, MAX_N_JETS = 6):
+    import os
+    include = os.path.join(fastforest_path, 'include')
+    lib = os.path.join(fastforest_path, 'lib')
+    ROOT.gSystem.AddIncludePath(f"-I{include}")
+    ROOT.gSystem.AddLinkedLibs(f"-L{lib} -lfastforest")
+    ROOT.gSystem.Load(f"{lib}/libfastforest.so.1")
+    ROOT.gSystem.CompileMacro("ml_helpers.cpp", "kO")
+
+    ROOT.gInterpreter.Declare('''
+                              auto models = get_fastforests("models/", 20);
+                              auto feven = models["even"];
+                              auto fodd = models["odd"];
+                              ''')
+
+    ROOT.gInterpreter.Declare(
+        f'''
+        size_t MAX_N_JETS = {MAX_N_JETS};
+        std::map<int, std::vector<ROOT::RVecI>> permutations = get_permutations_dict(MAX_N_JETS);
+        '''
+    )
+
+def prepare (df: ROOT.RDataFrame)->ROOT.RDataFrame:
+
+    
+    # prepare lepton fields (p4, eta, phi)
+    return (
+
+        df.Define('Electron_phi_cut', 'Electron_phi[Electron_mask]')
+          .Define('Electron_eta_cut', 'Electron_eta[Electron_mask]')
+          .Define(
+            'Electron_p4', 
+            '''
+            ConstructP4 (
+                    Electron_pt[Electron_mask], 
+                    Electron_eta_cut, 
+                    Electron_phi_cut, 
+                    Electron_mass[Electron_mask]
+            )
+            '''
+          )
+
+          .Define('Muon_phi_cut', 'Muon_phi[Muon_mask]')
+          .Define('Muon_eta_cut', 'Muon_eta[Muon_mask]')
+          .Define(
+            'Muon_p4', 
+            '''
+            ConstructP4 (
+                    Muon_pt[Muon_mask], 
+                    Muon_eta_cut, 
+                    Muon_phi_cut, 
+                    Muon_mass[Muon_mask]
+            )            
+            ''' 
+          )
+
+          .Define('Lepton_phi', 'Concatenate(Electron_phi_cut, Muon_phi_cut)')
+          .Define('Lepton_eta', 'Concatenate(Electron_eta_cut, Muon_eta_cut)')
+          .Define('Lepton_p4', 'Concatenate(Electron_p4, Muon_p4)')
+
+    )
+
+def define_features (df: ROOT.RDataFrame)->ROOT.RDataFrame:
+
+    
+    # prepare lepton fields (p4, eta, phi)
+    df2 = prepare(df)
+
+    # get indexes of four jets
+    df2 = (
+        df2.Define(
```

Comment:
should pick one: either increase X in `dfX` with every sequence of operations, or always re-assign to `df`